apps/src/aichat/redux/slice.ts

```diff
@@ -0,0 +1,286 @@
+import {createSlice, PayloadAction} from '@reduxjs/toolkit';
+
+import {registerReducers} from '@cdo/apps/redux';
+
+import {ModalTypes} from '../constants';
+import {
+  AiCustomizations,
+  ChatEvent,
+  ChatMessage,
+  FieldVisibilities,
+  LevelAichatSettings,
+  ModelCardInfo,
+  SaveType,
+  ViewMode,
+  Visibility,
+  isModelUpdate,
+  isNotification,
+  isChatMessage,
+  FeedbackValue,
+} from '../types';
+import {
+  DEFAULT_VISIBILITIES,
+  EMPTY_AI_CUSTOMIZATIONS,
+} from '../views/modelCustomization/constants';
+import {validateModelId} from '../views/modelCustomization/utils';
+
+const messageListKeys = ['chatEventsPast', 'chatEventsCurrent'] as const;
+type MessageLocation = {
+  index: number;
+  messageListKey: (typeof messageListKeys)[number];
+};
+type ModalType = ModalTypes.WARNING | ModalTypes.TEACHER_ONBOARDING | undefined;
+
+export interface AichatState {
```

Comment:
Because of the size of this file and its name, I think it may make sense to storing state in a separate file - but not a strong feeling!

Reply:
Makes sense! I may actually pull this and other local type (ModalType, etc) into a separate `types` file?

Reply:
Updated! Actually ended up slimming this down a bit more
- Got rid of the `MessageLocation` type since it can be inferred from the return type of `getUpdateMessageLocation`
- Got rid of `ModalType` since it's equivalent to `ModalTypes | undefined`
- Which left only the `AichatState` type. Moved this into a separate file called `state.ts`